packages/ContentAutotagging/Engine/generic/AutotagBaseEngine.ts

```diff
@@ -0,0 +1,745 @@
+        processRunParams.sourceID = contentItems[0].Get('ContentSourceID');
+        processRunParams.startTime = new Date();
+        processRunParams.numItemsProcessed = contentItems.length;
+
+        for (const contentItem of contentItems) {
+            try {
+                const processingParams = new ContentItemProcessParams();
+                
+                // Parameters that depend on the content item
+                processingParams.text = contentItem.Get('Text');
+                processingParams.contentSourceTypeID = contentItem.Get('ContentSourceTypeID');
+                processingParams.contentFileTypeID = contentItem.Get('ContentFileTypeID');
+                processingParams.contentTypeID = contentItem.Get('ContentTypeID');
+
+                // Parameters that depend on the content type
+                const { modelID, minTags, maxTags } = await this.getContentItemParams(processingParams.contentTypeID, contextUser) 
+                processingParams.modelID = modelID;
+                processingParams.minTags = minTags;
+                processingParams.maxTags = maxTags;
+                processingParams.contentItemID = contentItem.Get('ID');
+
+                this.ProcessContentItemText(processingParams, contextUser);
+
+            }
+
+            catch (e) {
+                console.error(`Failed to process content source item: ${contentItem.Get('contentItemID')}`);
+                throw e;
+            }
+        }
+
+        processRunParams.endTime = new Date();
+        await this.saveProcessRun(processRunParams, contextUser);
+    }
+
+    /**
+     * Given processing parameters that include the text from our content item, process the text with the LLM and extract the 
+     * information related to that content type.
+     * @param params 
+     * @returns 
+     */
+    public async ProcessContentItemText(params: ContentItemProcessParams, contextUser: UserInfo): Promise<void> {
+        const LLMResults: JsonObject = await this.promptAndRetrieveResultsFromLLM(params, contextUser);
+        await this.saveLLMResults(LLMResults, contextUser);
+    }
+
+    public async promptAndRetrieveResultsFromLLM(params: ContentItemProcessParams, contextUser: UserInfo) { 
+        const LLM = await this.getLLMClient(params.modelID, contextUser)
+        const model = await this.getModelAPINameFromID(params.modelID, contextUser)
+        const text = this.chunkExtractedText(params.text, model)
+        const contentType = await this.getContentTypeName(params.contentTypeID, contextUser)
+        const contentSourceType = await this.getContentSourceTypeName(params.contentSourceTypeID, contextUser)
+        const additionalContentTypePrompts = await this.getAdditionalContentTypePrompt(params.contentTypeID, contextUser)
+        const minTags = params.minTags
+        const maxTags = params.maxTags
+        const LLMResults: JsonObject = {}
+        const startTime = new Date()
+
+        for (const chunk of text) {
```

Comment:
@nico-oz we can ship `2.4` without this, but as soon as you get back from vacation you should parallelize this by running each internal component of the for loop with a promise and then do promise.all at the end of the loop so you do all of these in parallel. Also, this is a good example where functional decomposition is key to make your code more readable, break out the internal work of the for loop into its own method and call it from the for loop, that will make it more readable/maintainable. RULE OF THUMB: you have to be able read the function on a single page without scrolling, otherwise break it up into smaller functions.

Reply:
@nico-oz  make an issue for this in GitHub to fix this post-vacay so you don't lose track of it after this PR is closed pls.